### PR TITLE
Add all projects to MSBuildAllProjects property

### LINKED ISSUE #277: Make sure all projects add themselves to MSBuildAllProjects
Reminded by #276 that we're not doing this everywhere.

### PR DESCRIPTION
**Pull request by dsplaisted**
Fixes #277 

@nguerrera

### CODE REVIEW

**Review by rainersigwald — APPROVED** [review]

**Comment by nguerrera** [comment]
Tests are failing. I seem to be reviewing a lot of PRs in that state. Are you running build.cmd before sending these out for review?

**Review by nguerrera — CHANGES_REQUESTED** [review]
Waiting for green run before reviewing.

**Comment by dsplaisted** [comment]
> Tests are failing. I seem to be reviewing a lot of PRs in that state. Are you running build.cmd before sending these out for review?

I'm not sure I've always remembered to do so, but generally, yes.  These changes passed locally for me.

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs [hunk 1]
```diff
@@ -70,6 +70,78 @@ public void It_builds_the_library_twice_in_a_row()
...
+            string dotnetRoot = Path.GetDirectoryName(RepoInfo.DotNetHostPath);
+
+            expectedAllProjects = expectedAllProjects.Distinct().ToList();
+
+            var expectedBuiltinProjects = expectedAllProjects.Where(project => project.StartsWith(dotnetRoot, StringComparison.OrdinalIgnoreCase)).ToList();
+            var expectedIntermediateProjects = expectedAllProjects.Where(project => project.StartsWith(baseIntermediateDirectory, StringComparison.OrdinalIgnoreCase)).ToList();
+            var expectedOtherProjects = expectedAllProjects
+                .Except(expectedBuiltinProjects)
+                .Except(expectedIntermediateProjects)
+                //  TODO: Remove this when https://github.com/NuGet/Home/issues/3851 is fixed
```
Comment (nguerrera) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs]: File bug on SDK to represent the TODO and link to the blocking external issue.

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs [hunk 2]
```diff
@@ -70,6 +70,78 @@ public void It_builds_the_library_twice_in_a_row()
...
+            var otherProjectsFromProperty = allProjectsFromProperty.Except(builtinProjectsFromProperty).Except(intermediateProjectsFromProperty).ToList();
+
+            otherProjectsFromProperty.Should().BeEquivalentTo(expectedOtherProjects);
+
+            //  TODO: Uncomment the following lines when the following bugs are fixed:
+            //          - https://github.com/Microsoft/msbuild/issues/1298
+            //          - https://github.com/NuGet/Home/issues/3851
+            //          - https://github.com/dotnet/cli/issues/4571,
+            //          - https://github.com/dotnet/roslyn/issues/14870
+
```
Comment (nguerrera) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs]: Ditto.

**Review by nguerrera — APPROVED** [review]

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs (hunk 1 above)
Comment (dsplaisted) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs]: > File bug on SDK to represent the TODO and link to the blocking external issue.

To clarify, you mean file a bug in the SDK repo to remove the workaround in tests once the NuGet bug is fixed.  Correct?

Comment (nguerrera) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs]: Correct

Comment (dsplaisted) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildALibrary.cs]: OK, filed #380 for this

**Pull request merged**